velox/substrait/VeloxToSubstraitType.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/substrait/VeloxToSubstraitType.h"
+
+#include "velox/expression/Expr.h"
+
+namespace facebook::velox::substrait {
+
+void VeloxToSubstraitTypeConvertor::toSubstraitType(
+    const velox::TypePtr& vType,
+    ::substrait::Type* sType) {
+  switch (vType->kind()) {
+    case velox::TypeKind::BOOLEAN: {
+      sType->set_allocated_bool_(new ::substrait::Type_Boolean());
+      break;
+    }
+    case velox::TypeKind::TINYINT: {
+      sType->set_allocated_i8(new ::substrait::Type_I8());
+      break;
+    }
+    case velox::TypeKind::SMALLINT: {
+      sType->set_allocated_i16(new ::substrait::Type_I16());
+      break;
+    }
+    case velox::TypeKind::INTEGER: {
+      sType->set_allocated_i32(new ::substrait::Type_I32());
+      break;
+    }
+    case velox::TypeKind::BIGINT: {
+      sType->set_allocated_i64(new ::substrait::Type_I64());
+      break;
+    }
+    case velox::TypeKind::REAL: {
+      sType->set_allocated_fp32(new ::substrait::Type_FP32());
+      break;
+    }
+    case velox::TypeKind::DOUBLE: {
+      sType->set_allocated_fp64(new ::substrait::Type_FP64());
+      break;
+    }
+    case velox::TypeKind::VARCHAR: {
+      sType->set_allocated_varchar(new ::substrait::Type_VarChar());
+      break;
+    }
+    case velox::TypeKind::VARBINARY: {
+      sType->set_allocated_binary(new ::substrait::Type_Binary());
+      break;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      sType->set_allocated_timestamp(new ::substrait::Type_Timestamp());
+      break;
+    }
+    case velox::TypeKind::ARRAY: {
+      google::protobuf::Arena* arena;
```

Comment:
Hmm... how does this work. arena is an uninitialized pointer. arena->Own(sTList) would probably crash.

I think you need to create an Arena that would be used to allocate all objects in a given plan.

Reply:
how about just delete the sTList  `delete sTList;` when catch the exception. the mutable_type and set_allocated_list already consider the different situation.

 void Type::set_allocated_list(::substrait::Type_List* list) {
   ::google::protobuf::Arena* message_arena = GetArenaNoVirtual();
  clear_kind();
  if (list) {
    ::google::protobuf::Arena* submessage_arena = NULL;
    if (message_arena != submessage_arena) {
      list = ::google::protobuf::internal::GetOwnedMessage(
          message_arena, list, submessage_arena);
    }
    set_has_list();
    kind_.list_ = list;
  }
  // @@protoc_insertion_point(field_set_allocated:substrait.Type.list)
 }

Reply:
> how about just delete the sTList, `delete sTList;` when catch the exception

This pattern is fragile. It is easy to forget to add try-catch and leak memory. Is the code snippet above generated or hand-written code?

> the mutable_type and set_allocated_list already consider the different situation.

I'm not sure I understand. Would you explain?

Reply:
the code snippet above is generated by proto.